dotnet-engine/UnleashEngine/UnleashEngine.cs

```diff
@@ -0,0 +1,103 @@
+﻿
+using System.Runtime.InteropServices;
+using System.Text.Json;
+
+
+namespace Unleash;
+
+public class Context
+{
+    public string? UserId { get; set; }
+    public string? SessionId { get; set; }
+    public string? RemoteAddress { get; set; }
+    public string? Environment { get; set; }
+    public string? AppName { get; set; }
+    public string? CurrentTime { get; set; }
+    public Dictionary<string, string>? Properties { get; set; }
+}
+
+public class Variant
+{
+    public string Name { get; set; }
+    public Payload? Payload { get; set; }
+    public bool Enabled { get; set; }
+}
+
+public class Payload
+{
+    public string? PayloadType { get; set; }
+    public string? Value { get; set; }
+}
+
+
+public class UnleashEngine
+{
+    private JsonSerializerOptions options = new JsonSerializerOptions
+    {
+        PropertyNamingPolicy = JsonNamingPolicy.CamelCase
+    };
+
+    private IntPtr state;
+    private const string DLL_PATH = "../../../../../target/release/libyggdrasilffi.so";
+
+    [DllImport(DLL_PATH)]
+    private static extern IntPtr engine_new();
+
+    [DllImport(DLL_PATH)]
+    private static extern void engine_free(IntPtr ptr);
+
+    [DllImport(DLL_PATH)]
+    private static extern IntPtr engine_take_state(IntPtr ptr, string json);
+
+    [DllImport(DLL_PATH)]
+    private static extern bool engine_is_enabled(IntPtr ptr, string toggle_name, string context);
+
+    [DllImport(DLL_PATH)]
+    private static extern IntPtr engine_get_variant(IntPtr ptr, string toggle_name, string context);
+
+    [DllImport(DLL_PATH)]
+    private static extern void engine_free_variant_def(IntPtr ptr);
+
+    public UnleashEngine()
+    {
+        this.state = engine_new();
+    }
+
+    public void Dispose()
+    {
+        engine_free(this.state);
+        GC.SuppressFinalize(this);
+    }
+
+    public void TakeState(string json)
```

Comment:
So this is how you prep the engine behind the scenes, when you've fetched the togglestate from Unleash you feed it through here and then you can keep invoking from IsEnabled etc?

How is this state kept? Once per instance? Or is it per thread/process?

Reply:
This is a great question! Unfortunately the answer is complex and it requires some understanding of how the memory is allocated.

State is handled Rust side and the host language gets a pointer to that state. That's allocated by Rust's allocator, not the .NET allocator. So state is once per pointer returned by the constructor in the Rust code. In practice that's as many times as you want per thread (singleton stuff needs to be related to the toggle fetcher not this code) and memory between Rust and host language is handled by OS memory mapper so processes are isolated but in sync across Rust and host language